src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.Unix.cs

```diff
@@ -0,0 +1,150 @@
+                UnixFileMode.OtherRead | UnixFileMode.OtherWrite |  UnixFileMode.OtherExecute;
+
+            string filename = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+            FileStreamOptions options = new()
+            {
+                Mode = FileMode.CreateNew,
+                UnixCreateMode = OwnershipPermissions,
+                Options = FileOptions.DeleteOnClose,
+                Access = FileAccess.Write,
+                BufferSize = 0
+            };
+            using var fs = new FileStream(filename, options);
+            UnixFileMode actual = File.GetUnixFileMode(fs.SafeFileHandle);
+
+            return OwnershipPermissions & ~actual;
+        }
+
+        private sealed class ReverseStringComparer : IComparer<string>
+        {
+            public int Compare (string? x, string? y)
+                => StringComparer.Ordinal.Compare(y, x);
+        }
+
+        private static readonly ReverseStringComparer s_reverseStringComparer = new();
+
+        private static UnixFileMode UMask => s_umask.Value;
+
+        /*
+            Tar files are usually ordered: parent directories come before their child entries.
+
+            They may be unordered. In that case we need to create parent directories before
+            we know the proper permissions for these directories.
+
+            We create these directories with restrictive permissions. If we encounter an entry for
+            the directory later, we store the mode to apply it later.
+
+            If the archive doesn't have an entry for the parent directory, we use the default mask.
+
+            The pending modes to be applied are tracked through a reverse-sorted dictionary.
+            The reverse order is needed to apply permissions to children before their parent.
+            Otherwise we may apply a restrictive mask to the parent, that prevents us from
+            changing a child.
+        */
+
+        internal static SortedDictionary<string, UnixFileMode>? CreatePendingModesDictionary()
+            => new SortedDictionary<string, UnixFileMode>(s_reverseStringComparer);
+
+        internal static void CreateDirectory(string fullPath, UnixFileMode? mode, bool overwriteMetadata, SortedDictionary<string, UnixFileMode>? pendingModes)
+        {
+            // Restrictive mask for creating the missing parent directories while extracting.
+            const UnixFileMode ExtractPermissions = UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.UserExecute;
+
+            Debug.Assert(pendingModes is not null);
+
+            if (Directory.Exists(fullPath))
+            {
+                // Apply permissions to an existing directory when we're overwriting metadata
+                // or the directory was created as a missing parent (stored in pendingModes).
+                if (mode.HasValue)
+                {
+                    bool hasExtractPermissions = (mode & ExtractPermissions) == ExtractPermissions;
```

Comment:
Just like in lines 90 and 95, the other locations consuming the value of `mode` should call `mode.Value`:

```suggestion
                    bool hasExtractPermissions = (mode.Value & ExtractPermissions) == ExtractPermissions;
```